### DISCUSSION TITLE
How do I compile a custom .deb package?

### DISCUSSION THREAD

**Discussion by davehouser1** [strongswan/strongswan#1217] [Q&A]
Hi.

I need to compile a custom .deb package of Strongswan for Ubuntu 20.
I already know how to create a custom .rpm package for RHEL 8.4, and it works. 
Specifically I need to custom build the package because when I build a custom .rpm I enabled `--with-group=customgroup` in the `strongswan.spec` file. I will need to do the same to create the custom .deb file. 

The package will be used for offline installs. 
I searched through the install documentation but did not find anything about what I am trying to do. 
Do not want to install `alien` to install .rpms on Ubuntu. Want to use .deb packages and install with `dpkg` or `apt`

How can I build a custom .deb package that enables `--with-group` configuration?

**Comment by Thermi** [comment]
you can specify the group in strongswan.conf already

Other than that, take the source code of the package itself and modify it accordingly.
Package sources are available online, or via apt-get I suppose.

**Reply by Thermi** [reply]
Like here https://packages.debian.org/sid/strongswan

**Reply by davehouser1** [reply]
Hello @Thermi, interesting, your referring to `group` parameter in [strongswan.conf](https://docs.strongswan.org/docs/5.9/config/strongswanConf.html)? To confirm, if I set this value to `customgroup`, then restart strongswan.service, then any user in the `customgroup` will be able to use `swanctl` just like root? If this is true there is no real reason for me to compile a custom package of Strongswan aside from convince? Has this setting always been in Strongswan or is it something new that was added?

**Reply by davehouser1** [reply]
If the above is the case, then I do not need to create a custom package, but could just use a standard install 
However I am confused on how or where I can get a strongswan .deb file for install. [This page](https://docs.strongswan.org/docs/5.9/install/install.html#Distribution-packages) states there are packages for Ubuntu. When going to the Ubuntu site, there are links to each distro [https://packages.ubuntu.com/search?keywords=strongswan&searchon=names&suite=all&section=all](url), however when selecting the distro I am using (20.04), there does not appear to be a .deb file, but only source code tar balls (I think). 

Is there a way to get a .deb file or create my own? Is it possible to create an install package for Ubuntu without having to compile from source?

**Reply by davehouser1** [reply]
BTW, I just performed the following install to test the `group` parameter first.
```
apt-get install strongswan strongswan-pki libcharon-extra-plugins libcharon-extauth-plugins libstrongswan-extra-plugins strongswan-swanctl -y
```
I then adjusted strongswan.conf to the following:

```
charon {
        group = customgroup
        load_modular = yes
        make_before_break = yes
        install_routes = no
        plugins {
                include strongswan.d/charon/*.conf
        }
        retransmit_tries = 1
    # Only enable this if you want to see more low level logs of what strongswan is doing
    #filelog {
    #    charon{
    #        path = /var/log/charon.log
    #        default = 4
    #    }
    #}
}

include strongswan.d/*.conf
```

Confirmed my user is in that group

```
$ id
uid=1001(customgroup) gid=1000(customgroup) groups=1000(customgroup)
```

and restarted strongswan service

```
# systemctl restart strongswan-starter.service
```

The service is active. But when I try to run a simple swanctl command as customgroup I get an error. This does not happen with root and I get the help output. 

```
$ swanctl -h
abort initialization due to invalid configuration
```

Is this expected?

**Reply by Thermi** [reply]
swanctl needs to be able to read the swanctl config files and access the subdirectories. Can it do that?

**Reply by tobiasbrunner** [reply]
Thermi is right about that (i.e. the `/etc/swanctl` dir and its subdirs and files have to be readable by the configured group). But note that this particular error is caused if strongswan.conf is not readable, so make sure that's allowed for that group too.

**Reply by davehouser1** [reply]
@Thermi good point. I did not need to do this on my RHEL installs when enabling the `--with-group` option during config of the RPM, I guess I need to do it here. 

I made this script to make the permission changes

```
ipsec_dir="/etc/ipsec.d"
swanctl_dir="/etc/swanctl"
ss_dir="/etc/strongswan.d"

conf_ipsec="ipsec.conf"
conf_swanctl="swanctl.conf"
conf_strongswan="strongswan.conf"
group="customgroup"

sudo cp ipsec.conf /etc/$conf_ipsec
sudo cp strongswan.conf /etc/$conf_strongswan
sudo cp swanctl.conf /etc/swanctl/$conf_swanctl
chown -R root:$group $conf_strongswan
chown -R root:$group $conf_ipsec
chown -R root:$group $ipsec_dir
chown -R root:$group $swanctl_dir
chown -R root:$group $ss_dir
```

After restarting `strongswan-starter.service`, and trying to run `swanctl -q` I see the following errors

```
$ swanctl -q
agent plugin requires CAP_DAC_OVERRIDE capability
plugin 'agent': failed to load - agent_plugin_create returned NULL
connecting to 'unix:///var/run/charon.vici' failed: Permission denied
Error: connecting to 'default' URI failed: Permission denied
strongSwan 5.8.2 swanctl
usage:
  swanctl --load-all [--raw|--pretty] [--clear] [--noprompt]
           --help            (-h)  show usage information
           --clear           (-c)  clear previously loaded credentials
           --noprompt        (-n)  do not prompt for passwords
           --raw             (-r)  dump raw response message
           --pretty          (-P)  dump raw response message in pretty print
           --file            (-f)  custom path to swanctl.conf
           --debug           (-v)  set debug level, default: 1
           --options         (-+)  read command line options from file
           --uri             (-u)  service URI to connect to
```

Searching the Strongswan mail archive I found[ this link ](https://www.mail-archive.com/users@lists.strongswan.org/msg14939.html) where someone else had a similar issue. Does not appear to be a solution for this post. 

Any recommendations?

**Reply by tobiasbrunner** [reply]
The _agent_ plugin isn't relevant (you can disable the plugin if you want to avoid the message). But `swanctl` needs access to the vici socket. If the user doesn't have access to `/var/run` or the socket file isn't owned by a group the user is in, this won't work (the latter should be correct if `charon.group` is set, though). To avoid having to change the permissions of `/var/run`, moving the socket to e.g. `/etc/swanctl` might be an option (see [here](https://docs.strongswan.org/docs/5.9/plugins/vici.html#_configuration)).

**Reply by davehouser1** [reply]
Ok I am getting closer. I figure out how to download the ubuntu deb packages offline, adjust the `rules` file to include `--with-group=customgroup`, then compile them all into .deb files and install offline. 
This appears to have worked. I am going to share detailed instructions on all the steps I took to get it to work.
Before I do, I do want to fix the `agent` plugin. 
Would rather it work and not disable, is there a way to fix it? 
And worst case, I can disable it, how do I do that?  The Strongswan documentation states that disabling a plugin should be done during configuration time. I checked the [readme](https://docs.strongswan.org/docs/5.9/install/autoconf.html#_disable_options) but I dont see one for --enable-agent. I assume I need to remove it? I tried this and I get errors when compiling now 

```
# remove all .la files
find debian/tmp/usr/lib -name '*.la' -delete
# first special cases
# handle Linux-only plugins
dh_install -p libstrongswan-standard-plugins usr/lib/ipsec/plugins/libstrongswan-connmark.so
dh_install: warning: Cannot find (any matches for) "usr/lib/ipsec/plugins/libstrongswan-agent.so" (tried in ., debian/tmp)

dh_install: warning: libstrongswan-standard-plugins missing files: usr/lib/ipsec/plugins/libstrongswan-agent.so
dh_install: warning: Cannot find (any matches for) "usr/share/strongswan/templates/config/plugins/agent.conf" (tried in ., debian/tmp)

dh_install: warning: libstrongswan-standard-plugins missing files: usr/share/strongswan/templates/config/plugins/agent.conf
dh_install: warning: Cannot find (any matches for) "etc/strongswan.d/charon/agent.conf" (tried in ., debian/tmp)

dh_install: warning: libstrongswan-standard-plugins missing files: etc/strongswan.d/charon/agent.conf
dh_install: error: missing files, aborting
make[1]: *** [debian/rules:161: override_dh_install-arch] Error 255
make[1]: Leaving directory '/home/dh016d/projects/CDAP/strongswan_builds/apt_download/apt_download_03/strongswan-5.6.2'
make: *** [debian/rules:304: binary] Error 2
dpkg-buildpackage: error: fakeroot debian/rules binary subprocess returned exit status 2
```

I can only assume the plugin is apart of this shared library `dh_install -p libstrongswan-standard-plugins usr/lib/ipsec/plugins/libstrongswan-connmark.so`, its not clear to me how to disable this when compiling the ubuntu packages, any recommendations? 
Out of curiosity, what is the purpose of the agent plugin?

**Reply by tobiasbrunner** [reply]
> And worst case, I can disable it, how do I do that?

See [here](https://docs.strongswan.org/docs/5.9/plugins/pluginLoad.html).

> I assume I need to remove it?

All `--enable-*` options are not enabled by default. So just don't pass them to `configure` and they are "disabled" (explicitly passing the corresponding `--disable-*` option also works, just in case it will ever be enabled by default). The options that are documented as `--disable-*` are enabled by default, for those you'd have to explicitly use the option to disable the feature/plugin.

> I tried this and I get errors when compiling now

Yes, you have to adapt the appropriate `*.install` file(s).

>  Out of curiosity, what is the purpose of the agent plugin?

It allows authenticating via `ssh/gpg-agent`. It currently can only be used with `charon-cmd`  and the NetworkManager plugin. There currently isn't really a point in actually loading the plugin in `swanctl`.

**Reply by davehouser1** [reply]
> Yes, you have to adapt the appropriate *.install file(s).

Got it. I had to do the following
- Edit  strongswan-5.8.2/debian/rules. Remove `--enable-agent`.
- Edit strongswan-5.8.2/debian/libstrongswan-standard-plugins.install. Comment out 
    - usr/lib/ipsec/plugins/libstrongswan-agent.so
    - usr/share/strongswan/templates/config/plugins/agent.conf
    - etc/strongswan.d/charon/agent.conf
- Recompile.

Now I no longer get the agent errors when running it with a user in the `--with-group=customgroup`

Note: Trying to adjust `load=no` in /etc/strongswan.d/charon/agent.conf, while the agent plugin was compiled, this did not remove the error for any user in the `--with-group=customgroup` I added. Had to remove from compiling. 

Thanks @tobiasbrunner for your help!
I have some more testing to do with my process but will post here with the solution of the full process when complete.

**Reply by tobiasbrunner** [reply]
> Trying to adjust load=no in /etc/strongswan.d/charon/agent.conf, while the agent plugin was compiled, this did not remove the error for any user in the --with-group=customgroup I added

That's because that config snippet is only included in the `charon.plugins` section in `strongswan.conf` (as the `charon` sub-directory suggests). To disable it in `swanctl` you'd have to configure `swanctl.plugins.agent.load=no ` in `strongswan.conf`.

**Comment by cyanideangel2000** [comment]
I am seeing the same error on Ubuntu 20.04 even for the root user or with sudo even though as far as i can see the /etc/swanctl.conf exists and looks valid.
Strongswan version 5.8.2

**Reply by Thermi** [reply]
which error?

**Reply by cyanideangel2000** [reply]
here:
```
vtewari@host1:/opt/me/ipsecgw$ swanctl --version

abort initialization due to invalid configuration

vtewari@host1:/opt/me/ipsecgw$ swanctl -q

abort initialization due to invalid configuration

vtewari@host1:/opt/me/ipsecgw$ swanctl --help

abort initialization due to invalid configuration

vtewari@host1:/opt/me/ipsecgw$ ls -al
total 52

-rw-rw-r--  1 root    root      190 Aug 27 23:26 ipsec.conf

drwxrwxr-x 10 root    root     4096 Aug 28 01:10 ipsec.d

-rw-rw-r--  1 root    root      144 Aug 27 23:26 ipsec.secrets

-rw-rw-r--  1 root    root       26 Aug 27 23:26 netskope.default.ipsec.secrets

-rw-r--r--  1 root root 18943 Aug 18 15:06 default.swanctl.conf

-rw-r--r--  1 root    root     1634 Aug 28 02:23 strongswan.conf

-rw-r-----  1 root    root       55 Aug 28 02:20 swanctl.conf

vtewari@host1:/opt/me/ipsecgw$
```

I have set SYSCONFDIR to /opt/me/ispecgw and --with-swanctldir = /opt/me/ipsecgw in my configure
